### PR TITLE
Organize the directory structure of `JournalStorage`

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

To make it clearer, I organize the directory structure of `JournalStorage`.

## Description of the changes

- Rename some files and place each module in better locations

### CODE REVIEW

File: optuna/storages/_journal/file.py [hunk 1]
```diff
@@ -1,279 +1,27 @@
...
-            with open(self._file_path, "ab") as f:
-                f.write(what_to_write.encode("utf-8"))
-                f.flush()
-                os.fsync(f.fileno())
-
-
-@deprecated_class(
-    "4.0.0", "7.0.0", text="Use :class:`~optuna.storages.JournalFileBackend` instead."
+msg = _DEPRECATION_WARNING_TEMPLATE.format(
+    name="optuna.storages._journal.file.py", d_ver="4.0.0", r_ver="5.0.0"
```
Comment (nabenabe0928) [inline optuna/storages/_journal/file.py:14]:
```suggestion
    name="optuna.storages._journal.file.py", d_ver="4.0.0", r_ver="7.0.0"
```

**Comment by nabenabe0928** [comment]
TBD
- Whether to remove `_journal.file.py`

**Review by eukaryo — COMMENTED** [review]
Basically LGTM. However, I suppose adding unittest is needed.
We ought to have unittests that confirms FutureWarning is actually generated for all names and all imports that should be deprecated.

**Review by eukaryo — COMMENTED** [review]
I think the unittests should be done in separate PR,

**Comment by nabenabe0928** [comment]
Followup
- Deprecate the import of `optuna.storages.JournalFile(Symlink|Open)Lock` object from `optuna.storages` and recommend it from `optuna.storages.journal`
- Remove deprecated modules in `.journal.__init__.py`
- Add unit tests to check whether future warning is triggered when importing deprecated modules

**Review by eukaryo — APPROVED** [review]
LGTM. Thank you for your great effort! I expect your follow-up PR.

File: docs/source/reference/storages.rst [hunk 2]
```diff
@@ -13,9 +13,36 @@ The :mod:`~optuna.storages` module defines a :class:`~optuna.storages.BaseStorag
...
+   If users would like to use any backends not supported by Optuna, it is possible to do so by creating a customized class by inheriting :class:`optuna.storages.journal.BaseJournalBackend`.
+
+.. autosummary::
+   :toctree: generated/
+   :nosignatures:
+
+   optuna.storages.journal.JournalFileBackend
+   optuna.storages.journal.JournalRedisBackend
+
+Users can flexibly choose a lock object for backends and here is the list of supported lock objects:
```
Comment (not522) [inline docs/source/reference/storages.rst:29]:
```suggestion
Users can flexibly choose a lock object for :class:`~optuna.storages.journal.JournalFileBackend` and here is the list of supported lock objects:
```

File: optuna/storages/journal/__init__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from optuna.storages.journal._backend import JournalFileBackend
+from optuna.storages.journal._backend import JournalFileStorage
```
Comment (not522) [inline optuna/storages/journal/__init__.py:2]:
```suggestion
```

File: optuna/storages/journal/__init__.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+from optuna.storages.journal._base import BaseJournalBackend
+from optuna.storages.journal._base import BaseJournalLogStorage
```
Comment (not522) [inline optuna/storages/journal/__init__.py:4]:
```suggestion
```

File: optuna/storages/journal/__init__.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+from optuna.storages.journal._file_lock import JournalFileOpenLock
+from optuna.storages.journal._file_lock import JournalFileSymlinkLock
+from optuna.storages.journal._redis import JournalRedisBackend
+from optuna.storages.journal._redis import JournalRedisStorage
```
Comment (not522) [inline optuna/storages/journal/__init__.py:8]:
```suggestion
```

File: optuna/storages/journal/__init__.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+from optuna.storages.journal._storage import JournalStorage
+
+
+__all__ = [
+    "JournalFileBackend",
+    "JournalFileStorage",
```
Comment (not522) [inline optuna/storages/journal/__init__.py:14]:
```suggestion
```

File: optuna/storages/journal/__init__.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    "BaseJournalBackend",
+    "BaseJournalLogStorage",
```
Comment (not522) [inline optuna/storages/journal/__init__.py:16]:
```suggestion
```

File: optuna/storages/journal/__init__.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    "JournalFileOpenLock",
+    "JournalFileSymlinkLock",
+    "JournalRedisBackend",
+    "JournalRedisStorage",
```
Comment (not522) [inline optuna/storages/journal/__init__.py:20]:
```suggestion
```

**Review by not522 — COMMENTED** [review]
Basically LGTM. Could you check my comments?

File: optuna/storages/journal/__init__.py (hunk 3 above)
Comment (not522) [reply optuna/storages/journal/__init__.py:2]: They will be fixed by follow-up PRs.

**Review by not522 — APPROVED** [review]
LGTM!

**Pull request merged**